backend/lib/services/members.js

```diff
@@ -18,232 +18,497 @@
-  const body = {
-    spec: {
-      members
+
+  add (item) {
+    const id = item.id
+    if (this.has(id)) {
+      throw new TypeError('Subject already exists')
     }
+    this.set(id, item)
+  }
+
+  delete (id) {
+    delete this.subjectListItems[id]
+  }
+
+  has (id) {
+    return !!this.subjectListItems[id]
+  }
+
+  map (iteratee) {
+    return _.mapValues(this.subjectListItems, iteratee)
   }
-  return client['core.gardener.cloud'].projects.mergePatch(project.metadata.name, body)
 }
+exports.SubjectList = SubjectList
 
-// list, create and remove is done with the user
-exports.list = async function ({ user, namespace }) {
-  const client = user.client
+class ProjectMemberManager {
+  constructor (client, userId, project, serviceAccounts) {
+    this.client = client
+    this.userId = userId
+    this.project = project
+    this.serviceAccounts = serviceAccounts
+    this.subjectList = new SubjectList(project.spec.members)
+  }
 
-  // get project
-  const project = await readProject(client, namespace)
+  list () {
+    const namespace = this.project.spec.namespace
+    const ownServiceAccountMembers = members => _.filter(members, {
+      subject: {
+        kind: 'ServiceAccount',
+        namespace
+      }
+    })
 
-  // list serviceAccounts
-  const { items: serviceAccounts } = await client.core.serviceaccounts.list(namespace)
+    const assignServiceAccountMetadata = members => {
+      const serviceAccountMembers = ownServiceAccountMembers(members)
+      if (!_.isEmpty(serviceAccountMembers)) {
+        const metadataList = _
+          .chain(this.serviceAccounts)
+          .map('metadata')
+          .keyBy('name')
+          .value()
+        for (const member of serviceAccountMembers) {
+          const name = member.subject.name
+          member.createdBy = _.get(metadataList, [name, 'annotations', 'garden.sapcloud.io/createdBy'])
+          member.creationTimestamp = _.get(metadataList, [name, 'creationTimestamp'])
+        }
+      }
+      return members
+    }
 
-  // get project members from project
-  return fromResource(project, serviceAccounts)
-}
+    const projectMembers = this.subjectList.map('member')
+    const serviceAccountMembers = ownServiceAccountMembers(projectMembers)
+    const noMemberServiceAccounts = _
+      .chain(this.serviceAccounts)
+      .map('metadata')
+      .differenceBy(_.map(serviceAccountMembers, 'subject'), 'name')
+      .value()
 
-exports.get = async function ({ user, namespace, name }) {
-  const client = user.client
+    const noMemberServiceAccountSubjects = new SubjectList(_.map(noMemberServiceAccounts, ({ name, namespace }) => ({
+      name,
+      namespace,
+      kind: 'ServiceAccount'
+    })))
 
-  // get project
-  const project = await readProject(client, namespace)
+    const noMemberServiceAccountSubjectMembers = noMemberServiceAccountSubjects.map('member')
+    return _
+      .chain(projectMembers)
+      .concat(noMemberServiceAccountSubjectMembers)
```

Comment:
Da lag ich falsch. Es ist ja noch ein Object und kein Array.
```suggestion
      .assign(noMemberServiceAccountSubjectMembers)
```